hub/policy/src/main/java/uk/gov/ida/hub/policy/domain/controller/AbstractMatchRequestSentStateController.java

```diff
@@ -80,55 +79,83 @@ public void handleRequestFailure() {
+            LOG.error("Matching service request timed out for session {}", state.getSessionId());
+            MatchingServiceRequestErrorState matchingServiceRequestErrorState = new MatchingServiceRequestErrorState(
+                    state.getRequestId(),
+                    state.getRequestIssuerEntityId(),
+                    state.getSessionExpiryTimestamp(),
+                    state.getAssertionConsumerServiceUri(),
+                    state.getIdentityProviderEntityId(),
+                    state.getRelayState(),
+                    state.getSessionId(),
+                    state.getTransactionSupportsEidas());
+
+            stateTransitionAction.transitionTo(matchingServiceRequestErrorState);
+            return createResponseProcessingDetailsForStatus(ResponseProcessingStatus.SHOW_MATCHING_ERROR_PAGE);
+        }
+
+        return createResponseProcessingDetailsForStatus(ResponseProcessingStatus.WAIT);
+    }
+
+    @Override
+    public ResponseFromHub getErrorResponse() {
+        return responseFromHubFactory.createNoAuthnContextResponseFromHub(
+                state.getRequestId(),
+                state.getRelayState(),
+                state.getRequestIssuerEntityId(),
+                state.getAssertionConsumerServiceUri()
+        );
     }
 
     @Override
     public final void handleMatchResponseFromMatchingService(MatchFromMatchingService responseFromMatchingService) {
         validateResponse(responseFromMatchingService);
-
-        State nextState = getNextStateForMatch(responseFromMatchingService);
-
-        stateTransitionAction.transitionTo(nextState);
+        levelOfAssuranceValidator.validate(responseFromMatchingService.getLevelOfAssurance(), state.getIdpLevelOfAssurance());
+        transitionToNextStateForMatchResponse(responseFromMatchingService.getMatchingServiceAssertion());
     }
 
     @Override
     public final void handleNoMatchResponseFromMatchingService(NoMatchFromMatchingService responseFromMatchingService) {
         validateResponse(responseFromMatchingService);
-
-        State nextState = getNextStateForNoMatch();
-
-        stateTransitionAction.transitionTo(nextState);
+        transitionToNextStateForNoMatchResponse();
     }
 
-
     @Override
     public final void handleUserAccountCreatedResponseFromMatchingService(UserAccountCreatedFromMatchingService userAccountCreatedResponseFromMatchingService) {
         validateResponse(userAccountCreatedResponseFromMatchingService);
-
-        State nextState = getNextStateForUserAccountCreated(userAccountCreatedResponseFromMatchingService);
-
-        stateTransitionAction.transitionTo(nextState);
+        levelOfAssuranceValidator.validate(userAccountCreatedResponseFromMatchingService.getLevelOfAssurance(), state.getIdpLevelOfAssurance());
+        hubEventLogger.logUserAccountCreatedEvent(state.getSessionId(), state.getRequestIssuerEntityId(), state.getRequestId(), state.getSessionExpiryTimestamp());
```

Comment:
Why are we logging this now? Is it a must have? Do we know what impact it would be to event sink?

Reply:
Why wouldn't you want to log this?
Looking at the logic in other state controllers, we always log just before making a transition to a new state. And it makes sense to have a state transition trace in the logs for each session so you can investigate if something goes wrong. Also, the hubEventLogger already had the logUserAccountCreatedEvent method - as the name suggest, we should log at this point in the journey, I assumed.